### ISSUE TITLE
ReadWrite(): Sending tx.length bytes will be off by 1

### ISSUE THREAD

**Issue by sdunc** [swharden/FtdiSharp#10]
https://github.com/swharden/FtdiSharp/blob/20b3dbf50a740b9bc6995ca20a79dbba0614b47b/src/FtdiSharp/Protocols/SPI.cs#L170-L171

What FTDI calls LengthL and LengthH are more like indices.
A length of 0 contains a single byte. A length of 65535 contains 65536 bytes.
using tx.length directly will cause an extra byte of clock pulses to appear on spi bus.

From FT_000109 document:

![image (19)](https://github.com/user-attachments/assets/3fe809c5-7099-433c-8a61-7b7905bbe675)

![image](https://github.com/user-attachments/assets/29b2bf4e-aab1-4ba0-8aa9-bb5288409b95)

I believe replacing tx.length with tx.length - 1 in both lines will fix the issue.

best S

**Comment by swharden** [comment]
Hi @sdunc, thanks for the follow-up! I made the change suggested 21058cdd920745b16d9a2f10be091c843bf338c8 and am deploying FtdiSharp 0.1.1 to NuGet and it should be live in about 20 minutes.

If you give it a go, I'd love to hear if it worked as expected!

Thanks again
Scott

**Comment by sdunc** [comment]
Hey Scott,

0.1.0, note the extra byte on the right.
![image](https://github.com/user-attachments/assets/4e905486-67b9-4dde-a2ec-0bef374b01e4)

0.1.1, same source code. Correct number of bytes sent.
![image](https://github.com/user-attachments/assets/8dba7487-02e9-4a17-acee-b0e6d7b1eafe)

One thing you may want to improve would be chip select performance.
Your current approach:
https://github.com/swharden/FtdiSharp/blob/4cb31bc543d25660f6d18426294f5332a8bc2582/src/FtdiSharp/Protocols/SPI.cs#L90-L105

Calling FtdiDevice.Write() inside the loop incurs 5 transmission delays.
The approach used in the application note also sends the CS high and low commands 5x, but they fill a single buffer with all those bytes and .Write() only once with all the bytes. This change results in much much greater chip select performance.

With this change, note the improvement in how fast CS gets pulled back high:
![image](https://github.com/user-attachments/assets/c191a4f2-c413-4aa1-9cf4-39f368540f9c)

**Comment by swharden** [comment]
Hi @sdunc, thank you for this suggestion! 

> The approach used in the application note also sends the CS high and low commands 5x, but they fill a single buffer with all those bytes and .Write() only once with all the bytes. This change results in much much greater chip select performance.

Did you write a function I can copy/paste to replace the existing `CsHigh()` function? It you share that code, I would be happy to update it and publish a new release so it will be available for all. Thanks!

**Comment by sdunc** [comment]
Hey Scott, 
This is the unrolled loop. Even better performance can be achieved if the message to be sent as well as 5x CsLow() are added to the buffer for the same .Write() command. But that may require changing your interface. This is a good middle ground.

```cs
public void CsHigh() 
{ 
	// bit3:CS, bit2:MISO, bit1:MOSI, bit0:SCK 
    byte[] bytes = new byte[] 
    { 
        0x80,		// GPIO command for ADBUS 
        0b00001000, // value 
        0b00001011, // direction 
        0x80,
        0b00001000, 
        0b00001011,
        0x80,
        0b00001000,
        0b00001011,
        0x80,
        0b00001000,
        0b00001011,
        0x80,
        0b00001000,
        0b00001011,
    }; 
  
    if (!ClockIdlesLow) 
	{	 
     	bytes[1] |= 0b00000001;
     	bytes[4] |= 0b00000001;
     	bytes[7] |= 0b00000001;
     	bytes[10] |= 0b00000001;
     	bytes[13] |= 0b00000001;
	}
  
	FtdiDevice.Write(bytes); 
 }
```

**Comment by swharden** [comment]
Hi @sdunc, I just implemented this change for `CsHigh()` and `CsLow()` and put the new and old behaviors behind a feature flag to help debugging in case it doesn't work as expected.

I'm pushing version `0.1.2` to NuGet now and it should appear in about 10 minutes.

Are you able to glance at my implementation (and better yet give it a test with your hardware) to confirm it works as expected?

The code change I made is d7a1f581cbe39fdabe3298236c4830d62d1bdb03 but the important bit is here:

https://github.com/swharden/FtdiSharp/blob/3130a05a07489b2d4ab0dd4d19f235a151f72de1/src/FtdiSharp/Protocols/SPI.cs#L155-L182
